fx_compile.py

```diff
@@ -0,0 +1,432 @@
+                raise RuntimeError(f"Node referenced nonexistant target {'.'.join(target_atoms[:i])}")
+            attr_itr = getattr(attr_itr, atom)
+        return attr_itr
+
+    outs = None
+    inputs = []
+    module_attrs = []
+    for node in fx_model.graph.nodes:
+        if node.op == 'placeholder':
+            shapes = get_te_shapes(node.name)
+            env[node.name] = te.Placeholder(node.name, get_te_type(node.name), shapes)
+            inputs.append(env[node.name])
+        elif node.op == 'call_function':
+            result = lower_function(node.name, node.target, shape_env, lookup_env(node.args), node.args)
+            env[node.name] = result
+        elif node.op == 'output':
+            outs = list(lookup_env(node.args))
+        elif node.op == 'get_attr':
+            module_attrs.append(node)
+            env[node.name] = te.Placeholder(node.name, get_te_type(node.name), shapes)
+
+    loopnest = te.LoopNest(outs)
+    loopnest.prepare_for_codegen()
+    stmt = te.simplify(loopnest.root_stmt())
+    cg = te.construct_codegen('llvm', stmt, [te.BufferArg(x) for x in [env[i.name] for i in module_attrs] + inputs + outs])
+    def f(inps):
+        module_stuff = [fetch_attr(i.target) for i in module_attrs]
+        cg.call(module_stuff + list(inps))
+    return f
+
+class DeepAndWide(torch.nn.Module):
+    def __init__(self, num_features=50):
+        super(DeepAndWide, self).__init__()
+        self.mu = torch.nn.Parameter(torch.randn(1, num_features))
+        self.sigma = torch.nn.Parameter(torch.randn(1, num_features))
+        self.fc_w = nn.Parameter(torch.randn(1, num_features + 1))
+        self.fc_b = torch.nn.Parameter(torch.randn(1))
+
+    def forward(self, ad_emb_packed, user_emb, wide):
+        wide_offset = wide + self.mu
+        wide_normalized = wide_offset * self.sigma
+        wide_preproc = torch.clamp(wide_normalized, 0., 10.)
+        user_emb_t = torch.transpose(user_emb, 1, 2)
+        dp_unflatten = torch.bmm(ad_emb_packed, user_emb_t)
+        dp = torch.flatten(dp_unflatten, 1, -1)
+        inp = torch.cat([dp, wide_preproc], 1)
+        t1 = torch.transpose(self.fc_w, 1, 0)
+        fc1 = torch.addmm(self.fc_b, inp, t1)
+        return fc1
+
+
+with kernel_arena_scope():
+    with torch.no_grad():
+        num_features = 1000
+        mod = DeepAndWide(num_features)
+
+        # Phabricate sample inputs
+        batch_size = 5
+        embedding_size = 1000
+        ad_emb_packed = torch.randn(batch_size, 1, embedding_size)
+        user_emb = torch.randn(batch_size, 1, embedding_size)
+        wide = torch.randn(batch_size, num_features)
+        inps = (ad_emb_packed, user_emb, wide)
+        out = torch.empty(5, 1)
+
+        shape_env = ShapeProp(fx.symbolic_trace(mod)).propagate(*inps)
+        mod = decompose(mod, shape_env)
+        shape_env = ShapeProp(fx.symbolic_trace(mod)).propagate(*inps)
+        cg = nnc_compile(mod, True, shape_env)
```

Comment:
How bout `decompose` and `nnc_compile` do shape analysis internally to themselves? That way we're not symbolic tracing 4 times and adding more burden on the caller